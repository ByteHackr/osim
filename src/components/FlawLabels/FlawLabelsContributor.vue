<script setup lang="ts">
import { ref, toValue } from 'vue';

import { useMemoize, watchDebounced } from '@vueuse/core';

import { createCatchHandler } from '@/composables/service-helpers';
import { useFlaw } from '@/composables/useFlaw';

import { useUserStore } from '@/stores/UserStore';
import type { ZodJiraUserAssignableType } from '@/types';
import { searchJiraUsers } from '@/services/JiraService';
import DropDown from '@/widgets/DropDown/DropDown.vue';
import JiraUser from '@/widgets/JiraUser/JiraUser.vue';

const modelValue = defineModel<string>();

const contributor = ref<string>(modelValue?.value ?? '');
const isLoading = ref(false);
const results = ref<ZodJiraUserAssignableType[]>([]);

const { flaw } = useFlaw();
const userStore = useUserStore();

async function selfAssign() {
  await userStore.updateJiraUsername();
  modelValue.value = userStore.jiraUsername;
  contributor.value = userStore.jiraUsername;
}

const cacheJiraUsers = useMemoize(async (query: string) => {
  if (!query || !flaw.value.task_key || contributor.value === modelValue.value) {
    return [];
  }
  isLoading.value = true;
  const users = await searchJiraUsers(query, flaw.value.task_key)
    .catch(createCatchHandler('Failed to search jira users', false));
  isLoading.value = false;
  return users?.data?.users ?? [];
});

const onQueryChange = async (query: string) => {
  if (!query.length) {
    results.value = [];
    return;
  }

  const users = await cacheJiraUsers(query);
  results.value = users ?? [];
};

const handleSuggestionClick = (user: ZodJiraUserAssignableType) => {
  contributor.value = toValue(user.name);
  modelValue.value = contributor.value;
  results.value = [];
};

watchDebounced(contributor, onQueryChange, { debounce: 300 });
</script>
<template>
  <div class="input-group input-group-sm">
    <DropDown v-if="results.length">
      <div
        v-for="user in results"
        :key="user.name"
        class="item"
        @click.prevent.stop="handleSuggestionClick(user)"
      >

        <JiraUser v-bind="user" :query="contributor" />
      </div>
    </DropDown>
    <input
      v-model="contributor"
      type="text"
      class="form-control"
      :disabled="isLoading"
    >
    <button
      type="button"
      class="btn btn-secondary osim-self-assign"
      title="Self assign"
      :disabled="isLoading"
      @click="selfAssign"
    >
      <span v-osim-loading.grow="isLoading" class="self-assign" :class="{loading: isLoading}"></span>
    </button>
  </div>
</template>

<style scoped lang="scss">
.self-assign {
  display: inline-block;
  width: 12ch;
  text-align: center;

  &:not(.loading)::before {
    content: 'Self Assign';
  }
}
</style>
