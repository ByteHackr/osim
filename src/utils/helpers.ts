import { toRaw, isRef, isReactive, isProxy, ref, toRef, watch } from 'vue';
import * as R from 'ramda';

export function vRef(prop: Record<string, any>, property: string, defaultValue: any) {
  const reffedProp = toRef(prop, property);
  watch(reffedProp, (value) => flexRef.value = value);
  const flexRef = reffedProp.value === undefined ? ref(defaultValue) : reffedProp;
  return flexRef;
}

export function deepToRaw<T extends Record<string, any>>(sourceObj: T): T {
  const objectIterator = (input: any): any => {
    if (Array.isArray(input)) {
      return input.map((item) => objectIterator(item));
    }
    if (isRef(input) || isReactive(input) || isProxy(input)) {
      return objectIterator(toRaw(input));
    }
    if (input && typeof input === 'object') {
      return Object.keys(input).reduce((acc, key) => {
        acc[key as keyof typeof acc] = objectIterator(input[key]);
        return acc;
      }, {} as T);
    }
    return input;
  };

  return objectIterator(sourceObj);
}

export function deepCopyFromRaw<T extends Record<string, any>>(sourceObj: T): T {
  return JSON.parse(JSON.stringify(deepToRaw(sourceObj)));
}

// eslint-disable-next-line max-len
// https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string,
) =>
    array.reduce(
      (acc, value, index, array) => {
        (acc[predicate(value, index, array)] ||= []).push(value);
        return acc;
      },
    {} as { [key: string]: T[] },
    );

export const assignKeyValue = (object: Record<string, any>, key: string, value: any = null) => {
  object[key] = value;
  return object;
};

export const objectMap = (object: Record<string, any>, mapFn: (key: string, value: any) => any) =>
  Object.keys(object).reduce(
    (acc, key) => assignKeyValue(acc, key, mapFn(key, object[key])),
    {} as Record<string, any>,
  );

export const sortedByGroup = <T extends Record<string, any>>(array: T[], key: string) =>
  groupBy(
    array
      .filter((item: T) => item[key])
      .sort((itemA: T, itemB: T) => itemA[key].localeCompare(itemB[key])),
    (item: T) => item[key],
  );

type DeepMappable = any[] | Record<string, any>

const isNonEmptyArray = (value: any) => R.is(Array, value) && value.length > 0;
const isNonArrayObject = (value: any) => R.is(Object, value) && !R.is(Array, value);
const isDeepMappable = (value: DeepMappable) => isNonEmptyArray(value) || isNonArrayObject(value);

export const deepMap = (transform: (arg: any) => any, object: DeepMappable): any =>
  R.map(
    (val: any) => isDeepMappable(val)
      ? deepMap(transform, val)
      : transform(val)
    , object
  );

export const cveRegex = /^\s*(CVE-\d{4}-\d{4,7})\s*$/;
