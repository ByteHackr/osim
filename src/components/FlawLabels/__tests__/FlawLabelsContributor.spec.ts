import { ref } from 'vue';

import { flushPromises } from '@vue/test-utils';

import { mountWithConfig } from '@/__tests__/helpers';

import FlawLabelsContributor from '../FlawLabelsContributor.vue';

vi.mock('@/stores/UserStore', () => ({
  useUserStore: () => ({
    jiraUsername: 'skynet',
    updateJiraUsername: vi.fn(),
  }),
}));

vi.mock('@/services/JiraService', () => ({
  searchJiraUsers: () => {
    return Promise.resolve({ data: { users: [{ displayName: 'SkyNet', name: 'skynet' }] } });
  },
}));

vi.mock('@/composables/useFlaw', () => ({
  useFlaw: () => ({
    flaw: ref({
      task_key: '123',
    }),
  }),
}));

describe('flawLabelsContributor', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('should render', () => {
    const wrapper = mountWithConfig(FlawLabelsContributor);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should assign the contributor on self assign', async () => {
    const wrapper = mountWithConfig(FlawLabelsContributor, {
      props: {
        modelValue: 'skynet',
      },
    });

    await wrapper.find('button').trigger('click');

    expect(wrapper.props('modelValue')).toBe('skynet');
  });

  it('should assign the contributor on click', async () => {
    const wrapper = mountWithConfig(FlawLabelsContributor, {
      props: { 'onUpdate:modelValue': (e: string) => wrapper.setProps({ modelValue: e }) },
    });

    await wrapper.find('input').setValue('skynet');

    vi.runAllTimers();
    await flushPromises();
    await wrapper.find('div.item').trigger('click');

    expect(wrapper.props('modelValue')).toBe('skynet');
  });
});
