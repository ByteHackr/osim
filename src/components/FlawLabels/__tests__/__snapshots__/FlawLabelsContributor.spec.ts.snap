// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flawLabelsContributor > should render 1`] = `
"<div data-v-fd829dac="" class="input-group input-group-sm">
  <!--v-if--><input data-v-fd829dac="" type="text" class="form-control"><button data-v-fd829dac="" type="button" class="btn btn-secondary osim-self-assign" title="Self assign"><span data-v-fd829dac="" class="self-assign"></span></button>
</div>"
`;
